python/shotgun_data/background_task_manager.py

```
@@ -0,0 +1,832 @@
+        self._next_group_id += 1
+        return group_id
+
+    def _log(self, msg):
+        """
+        Wrapper method for logging useful information to debug.
+
+        :param msg: The message to be logged.
+        """
+        self._bundle.log_debug("Task Manager: %s" % msg)
+
+    def start_processing(self):
+        """
+        Start processing of tasks
+        """
+        self._can_process_tasks = True
+        self._start_tasks()
+
+    def pause_processing(self):
+        """
+        Pause processing of tasks - any currently running tasks will
+        complete as normal.
+        """
+        self._can_process_tasks = False
+        # and just let the current threads complete...
+
+    def shut_down(self):
+        """
+        Shut down the task manager.  This clears the task queue and gracefully stops all running
+        threads.  Completion/failure of any currently running tasks will be ignored.
+        """
+        self._log("Shutting down...")
+        self._can_process_tasks = False
+
+        # clear out all storage:
+        self._running_tasks = {}
+        self._available_threads = []
+        self._pending_tasks_by_priority = {}
+        self._tasks_by_id = {}
+        self._group_task_map = {}
+        self._upstream_task_map = {}
+        self._downstream_task_map = {}
+
+        # stop all worker threads and then wait for them to complete:
+        #for thread in self._all_threads:
+        #    #thread.stop(wait_for_completion = False)
+        #    thread.quit()
+        self._log("Waiting for %d background threads to stop..." % len(self._all_threads))
+        for thread in self._all_threads:
+            thread.shut_down()
+            #thread.deleteLater()
+            #thread = None
+            #thread.stop(wait_for_completion = True)
+        self._log("Shut down successfully!")
+        self._all_threads = []
+
+    def add_task(self, cbl, priority=None, group=None, upstream_task_ids=None, args=[], **kwargs):
```

The signature of this method can hide subtle bugs. Invoking the method like this `add_task(cb, upstream_taks_ids=[1])` will silently fail if the cb itself takes a kwargs as well or will fail later and not at the point of invocation of the method.